gst/datarepo/gstdatareposink.c

```diff
@@ -0,0 +1,466 @@
+{
+  GstDataRepoSink *sink = GST_DATA_REPO_SINK (object);
+
+  switch (prop_id) {
+    case PROP_LOCATION:
+      sink->filename = g_value_dup_string (value);
+      GST_INFO_OBJECT (sink, "filename = %s", sink->filename);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+  }
+}
+
+/**
+ * @brief Getter datareposink properties.
+ */
+static void
+gst_data_repo_sink_get_property (GObject * object, guint prop_id,
+    GValue * value, GParamSpec * pspec)
+{
+  GstDataRepoSink *sink;
+
+  sink = GST_DATA_REPO_SINK (object);
+
+  switch (prop_id) {
+    case PROP_LOCATION:
+      g_value_set_string (value, sink->filename);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+  }
+}
+
+/**
+ * @brief Called when a buffer should be presented or ouput.
+ */
+static GstFlowReturn
+gst_data_repo_sink_render (GstBaseSink * bsink, GstBuffer * buf)
+{
+  GstDataRepoSink *sink = GST_DATA_REPO_SINK_CAST (bsink);
+  gsize write_size = 0;
+  GstMapInfo info;
+  guint to_write = 0, byte_write = 0;
+
+  g_return_val_if_fail (sink->fd != 0, GST_FLOW_ERROR);
+
+  GST_OBJECT_LOCK (sink);
+  gst_buffer_map (buf, &info, GST_MAP_READ);
+  to_write = info.size;
+
+  GST_LOG_OBJECT (sink,
+      "Writing %d bytes at offset 0x%" G_GINT64_MODIFIER "x (%d size)",
+      to_write, sink->offset + byte_write, (guint) sink->offset + byte_write);
+
+  write_size = write (sink->fd, info.data, info.size);
+  if (write_size != info.size) {
+    GST_ERROR_OBJECT (sink, "Could not write data to file");
+    goto error;
+  }
+
+  gst_buffer_unmap (buf, &info);
+  GST_OBJECT_UNLOCK (sink);
+
+  sink->offset += write_size;
+
+  return GST_FLOW_OK;
+
+error:
+  gst_buffer_unmap (buf, &info);
+  GST_OBJECT_UNLOCK (sink);
+
+  return GST_FLOW_ERROR;
+}
+
+/**
+ * @brief Get caps of datareposink.
+ */
+static GstCaps *
+gst_data_repo_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
+{
+  GstDataRepoSink *sink = GST_DATA_REPO_SINK (bsink);
+  GstCaps *caps = NULL;
+
+  GST_OBJECT_LOCK (sink);
+  caps = gst_pad_get_pad_template_caps (GST_BASE_SINK_PAD (bsink));
+  GST_OBJECT_UNLOCK (sink);
+
+  GST_INFO_OBJECT (sink, "Got caps %" GST_PTR_FORMAT, caps);
+
+  if (caps && filter) {
+    GstCaps *intersection =
+        gst_caps_intersect_full (filter, caps, GST_CAPS_INTERSECT_FIRST);
+
+    gst_caps_unref (caps);
+    caps = intersection;
+  }
+
+  GST_DEBUG_OBJECT (sink, "result get caps: %" GST_PTR_FORMAT, caps);
+
+  return caps;
+}
+
+/**
+ * @brief Set caps of datareposink.
+ */
+static gboolean
+gst_data_repo_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
+{
+  GstDataRepoSink *sink;
+
+  sink = GST_DATA_REPO_SINK (bsink);
+  GST_INFO_OBJECT (sink, "set caps %" GST_PTR_FORMAT, caps);
+
+  return TRUE;
+}
+
+/**
+ * @brief Change state of datareposink.
+ */
+static GstStateChangeReturn
+gst_data_repo_sink_change_state (GstElement * element,
```

Comment:
If we do nothing while changing the state, we can remove this callback.
(Also, set-caps and get-caps would be unnecessary.)

Reply:
Thank you.
set-caps and get-caps is added for #4052.
I will remove change_state and I will add it feature if necessary.